src/map.rs

```diff
@@ -96,30 +185,31 @@ impl Map {
         );
 
         let infinite = infinite.unwrap_or(false);
-        let source_path = map_path.and_then(|p| p.parent());
 
-        let mut tilesets = Vec::new();
         let mut layers = Vec::new();
         let mut properties = HashMap::new();
+        let mut tilesets = Vec::new();
         parse_tag!(parser, "map", {
             "tileset" => |attrs| {
-                tilesets.push(Tileset::parse_xml(parser, attrs, source_path)?);
+                let res = Tileset::parse_xml_in_map(parser, attrs, map_path)?;
+                tileset_cache.get_or_insert(&res.tileset_path, res.tileset);
+                tilesets.push(TilesetRef{first_gid: res.first_gid, path: res.tileset_path});
```

Comment:
I think there are two problems here:

* The tileset, be it embedded or external, is still always parsed, even if it is already part of the cache.
* We insert embedded tilesets in the cache using an empty path.

I'm thinking `EmbeddedParseResult` should have an `Option<Tileset>` and an `Option<PathBuf>`, or we should use an enum to say it's either an embedded tileset, or a reference to an external one.

In case it's a reference, we should not immediately call `Tileset::new_external` from `parse_xml_reference`, but rather just return the `tileset_path`, such that we can obtain the tileset from the cache (or load it, if not present) in `Map::parse_xml`. To enable this, `TilesetCache::get_or_insert` could probably be changed to `get_or_insert_with` and take function.

Reply:
> To enable this, TilesetCache::get_or_insert could probably be changed to get_or_insert_with and take a function.

It's a bit more complex than that since `new_external` returns a Result, and we can't handle the result inside of the function we pass in. Hmm

Reply:
Fixed in latest commit